src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java

```diff
@@ -0,0 +1,62 @@
+package com.dev.cinema.dao.impl;
+
+import com.dev.cinema.dao.UserDao;
+import com.dev.cinema.exception.DataProcessingException;
+import com.dev.cinema.lib.Dao;
+import com.dev.cinema.model.User;
+import com.dev.cinema.util.HibernateUtil;
+import java.util.List;
+import java.util.Optional;
+import org.hibernate.Session;
+import org.hibernate.Transaction;
+import org.hibernate.query.Query;
+
+@Dao
+public class UserDaoImpl implements UserDao {
+    @Override
+    public User add(User user) {
+        Transaction transaction = null;
+        Session session = null;
+        try {
+            session = HibernateUtil.getSessionFactory().openSession();
+            transaction = session.beginTransaction();
+            session.persist(user);
+            transaction.commit();
+            return user;
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataProcessingException(
+                    "Cant add user with email: " + user.getEmail()
+                            + " to the database", e);
+        } finally {
+            if (session != null) {
+                session.close();
+            }
+        }
+    }
+
+    @Override
+    public List<User> getAll() {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            Query<User> getAllUsersQuery = session.createQuery(
+                    "FROM User", User.class);
+            return getAllUsersQuery.getResultList();
+        } catch (Exception e) {
+            throw new DataProcessingException("Cant get all users from the database", e);
+        }
+    }
+
+    @Override
+    public Optional<User> findByEmail(String email) {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            Query<User> findByEmailQuery = session.createQuery(
+                    "FROM User WHERE email = :userEmail", User.class)
+                    .setParameter("userEmail", email);
+            return Optional.ofNullable(findByEmailQuery.uniqueResult());
```

Comment:
we can use findByEmailQuery.uniqueResultOptional();

Reply:
done